src/ReactFinalForm.js

```diff
@@ -34,63 +36,128 @@ export const all: FormSubscription = formSubscriptionItems.reduce(
+    const f = createForm(config)
+    f.pauseValidation()
+    return f
+  })
+  const firstRender = React.useRef(true)
 
-  constructor(props: Props) {
-    super(props)
-    const {
-      children,
-      component,
-      render,
-      subscription,
-      decorators,
-      ...rest
-    } = props
-    const config: Config = rest
-    this.mounted = false
-    try {
-      this.form = createForm(config)
-    } catch (e) {
-      // istanbul ignore next
-      if (process.env.NODE_ENV !== 'production') {
-        console.error(`Warning: ${e.message}`)
-      }
-    }
-    this.unsubscriptions = []
-    if (this.form) {
-      // set initial state
+  // synchronously register and unregister to query form state for our subscription on first render
+  const [state, setState] = React.useState<FormState>(
+    (): FormState => {
       let initialState: FormState = {}
-      this.form.subscribe((state: FormState) => {
+      form.subscribe(state => {
         initialState = state
       }, subscription || all)()
-      this.state = { state: initialState }
-    }
-    if (decorators) {
-      decorators.forEach(decorator => {
-        this.unsubscriptions.push(decorator(this.form))
-      })
+      return initialState
     }
-  }
+  )
 
-  notify = (state: FormState) => {
-    if (this.mounted) {
-      this.setState({ state })
+  const flattenedSubscription = flattenSubscription(subscription || all)
+  React.useEffect(() => {
+    // We have rendered, so all fields are no registered, so we can unpause validation
+    form.isValidationPaused() && form.resumeValidation()
+    const unsubscriptions: Unsubscribe[] = [
+      form.subscribe(s => {
+        if (firstRender.current) {
+          if (validationStateChanged(state, s)) {
+            // this may happen if we have field-level validation
+            setState(s)
+          }
+        } else {
+          setState(s)
+        }
+        firstRender.current = false
+      }, subscription || all),
+      ...(decorators
+        ? decorators.map(decorator =>
+            // this noop ternary is to appease the flow gods
+            // istanbul ignore next
+            decorator(form)
+          )
+        : [])
+    ]
+
+    return () => {
+      unsubscriptions.forEach(unsubscribe => unsubscribe())
     }
-    this.mounted = true
-  }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [decorators, ...flattenedSubscription])
 
-  handleSubmit = (event: ?SyntheticEvent<HTMLFormElement>) => {
+  // warn about decorator changes
+  useWhenValueChanges(
+    decorators,
+    () => {
+      console.error(
+        'Warning: Form decorators should not change from one render to the next as new values will be ignored'
+      )
+    },
+    // this env check short circuts the comparison in production
+    // so we don't run the shallowEqual on every render
+    // istanbul ignore next
+    process.env.NODE_ENV === 'production' ? () => true : shallowEqual
```

Comment:
u could also wrap this `useWhenValueChanges` in env check to ensure that this string etc goes away correctly in production, you can break the hooks rules if u know what you are doing 😉

Reply:
GREAT point! 👍